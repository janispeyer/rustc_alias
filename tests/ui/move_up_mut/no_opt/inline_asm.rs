// compile-flags: -Zmir-emit-retag
use std::arch::asm;

fn inline_asm(x: &mut i32, y: &mut i32) -> i32 {
    *x = 42;
    unsafe {
        asm!(
            "mov {tmp}, 7",
            tmp = out(reg) _,
        );
    }
    *y = 7;
    return *x;
}

fn main() {
    let mut x = 1;
    let mut y = 2;
    inline_asm(&mut x, &mut y);
}
