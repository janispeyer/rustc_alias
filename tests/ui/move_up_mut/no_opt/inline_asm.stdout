# Analysing main
# Analysing inline_asm
# MaybeTopOfBorrowStack Analysis
bb0
[ 0] [_1, _2] <- Retag([fn entry] _1)
[ 1] [_1, _2] <- Retag([fn entry] _2)
[ 2] [_1, _2] <- (*_1) = const 42_i32
[ 3] [_1, _2] <- StorageLive(_4)
[ 4] [_1, _2] <- StorageLive(_5)
[ 5] [_1, _2] <- _5 = const ()
[ 6] [] <- asm!("mov {0}, 7", out(reg) _, options((empty))) -> bb1

bb1
[ 0] [] <- StorageDead(_5)
[ 1] [] <- _4 = const ()
[ 2] [] <- StorageDead(_4)
[ 3] [] <- (*_2) = const 7_i32
[ 4] [] <- _0 = (*_1)
[ 5] [] <- return

# FindImmutabilitySpans Analysis
bb0
[ 0] {} <- Retag([fn entry] _1)
[ 1] {} <- Retag([fn entry] _2)
[ 2] {_1: Span(bb0[2])} <- (*_1) = const 42_i32
[ 3] {_1: Span(bb0[2])} <- StorageLive(_4)
[ 4] {_1: Span(bb0[2])} <- StorageLive(_5)
[ 5] {_1: Span(bb0[2])} <- _5 = const ()
[ 6] {_1: Span(bb0[2])} <- asm!("mov {0}, 7", out(reg) _, options((empty))) -> bb1

bb1
[ 0] {_1: Top} <- StorageDead(_5)
[ 1] {_1: Top} <- _4 = const ()
[ 2] {_1: Top} <- StorageDead(_4)
[ 3] {_1: Top} <- (*_2) = const 7_i32
[ 4] {_1: Top} <- _0 = (*_1)
[ 5] {_1: Top} <- return


# CFG Before
bb0
Retag([fn entry] _1)
Retag([fn entry] _2)
(*_1) = const 42_i32
StorageLive(_4)
StorageLive(_5)
_5 = const ()
asm!("mov {0}, 7", out(reg) _, options((empty))) -> bb1

bb1
StorageDead(_5)
_4 = const ()
StorageDead(_4)
(*_2) = const 7_i32
_0 = (*_1)
return

# CFG After
bb0
Retag([fn entry] _1)
Retag([fn entry] _2)
_6 = const 42_i32
(*_1) = _6
StorageLive(_4)
StorageLive(_5)
_5 = const ()
asm!("mov {0}, 7", out(reg) _, options((empty))) -> bb1

bb1
StorageDead(_5)
_4 = const ()
StorageDead(_4)
(*_2) = const 7_i32
_0 = (*_1)
return

